vmsyscall/src/memory_map.rs

```diff
@@ -0,0 +1,335 @@
+                end_frame_number: 0,
+            },
+            region_type: MemoryRegionType::Empty,
+        }
+    }
+}
+
+/// A range of frames with an exclusive upper bound.
+#[derive(Clone, Copy, PartialEq, Eq)]
+#[repr(C)]
+pub struct FrameRange {
+    /// The frame _number_ of the first 4KiB frame in the region.
+    ///
+    /// This convert this frame number to a physical address, multiply it with the
+    /// page size (4KiB).
+    pub start_frame_number: u64,
+    /// The frame _number_ of the first 4KiB frame that does no longer belong to the region.
+    ///
+    /// This convert this frame number to a physical address, multiply it with the
+    /// page size (4KiB).
+    pub end_frame_number: u64,
+}
+
+impl FrameRange {
+    /// Create a new FrameRange from the passed start_addr and end_addr.
+    ///
+    /// The end_addr is exclusive.
+    pub fn new(start_addr: u64, end_addr: u64) -> Self {
+        let last_byte = end_addr - 1;
+        FrameRange {
+            start_frame_number: start_addr / PAGE_SIZE,
+            end_frame_number: (last_byte / PAGE_SIZE) + 1,
+        }
+    }
+
+    /// Returns true if the frame range contains no frames.
+    pub fn is_empty(&self) -> bool {
+        self.start_frame_number == self.end_frame_number
+    }
+
+    /// Length of the frame range
+    pub fn len(&self) -> u64 {
+        self.end_frame_number - self.start_frame_number
+    }
+
+    /// Returns the physical start address of the memory region.
+    pub fn start_addr(&self) -> u64 {
+        self.start_frame_number * PAGE_SIZE
+    }
+
+    /// Returns the physical end address of the memory region.
+    pub fn end_addr(&self) -> u64 {
+        self.end_frame_number * PAGE_SIZE
+    }
+}
+
+impl fmt::Debug for FrameRange {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            f,
+            "FrameRange({:#x}..{:#x})",
+            self.start_addr(),
+            self.end_addr()
+        )
+    }
+}
+
+/// Represents possible types for memory regions.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+#[repr(C)]
+pub enum MemoryRegionType {
+    /// Unused memory, can be freely used by the kernel.
+    Usable,
+    /// Memory that is already in use.
+    InUse,
+    /// Memory reserved by the hardware. Not usable.
+    Reserved,
+    /// ACPI reclaimable memory
+    AcpiReclaimable,
+    /// ACPI NVS memory
+    AcpiNvs,
+    /// Area containing bad memory
+    BadMemory,
+    /// Memory used for loading the kernel.
+    Kernel,
+    /// Memory used for loading the ELF app.
+    App,
+    /// Memory used by the bootloader.
+    Bootloader,
+    /// Frame at address zero.
+    ///
+    /// (shouldn't be used because it's easy to make mistakes related to null pointers)
+    FrameZero,
+    /// An empty region with size 0
+    Empty,
+    /// Additional variant to ensure that we can add more variants in the future without
+    /// breaking backwards compatibility.
+    NonExhaustive,
```

Comment:
why did you remove `#[doc(hidden)]` ??

Reply:
No good reason. Just saw the comment above it and thought it'd be useful information to appear in the generated documentation.

Reply:
Is there a reason we can't use `#![non_exhaustive]` here?

Reply:
> Is there a reason we can't use `#![non_exhaustive]` here?

probably not